### PR TITLE
Fix VS issue with unhandled exception on secondary threads

### LINKED ISSUE #103385: Fully unhandled exceptions break debugger with empty stack
### Description

This is very similar to #101729, but in that issue, I wasn't clear if we were discussing user-unhandled exceptions (debugger stopping at the catch handler found) or fully unhandled exceptions. User-unhandled was addressed by the fix for #101729, but fully unhandled is still a problem.

### Reproduction Steps

Follow the same repro steps from #101729, but disable "Just My Code" before starting debugging

### Expected behavior

Debugger should break while there are still frames on the call stack

### Actual behavior

All frames are unwound when the debugger breaks

### Regression?

Regression from .NET 8

### Known Workarounds

Set the environment variable `DOTNET_LegacyExceptionHandling` to `1`

### Configuration

.NET 9 Preview 5 on Windows x64. I didn't try any other configuration.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by janvorli**
When "just my code" is disabled, unhandled exceptions on secondary threads cause the VS to stop without any stack trace shown. This is similar to the recently fixed problem that was happening with user unhandled exception treatment.
The exception is rethrown as native exception after leaving the last managed frame and it propagates to the `ManagedThreadBase_DispatchOuter` where it is caught. The debugger gets notified from there, but all of the managed stack frames are gone at that point, so the debugger cannot show them.
The fix is to report exception on a secondary thread as unhandled earlier, right in the SfiNext where we report it for the primary threads. The secondary thread is different in having the DebuggerU2MCatchHandlerFrame on stack while in the primary thread case, there is no explicit frame.

Close #103385

### CODE REVIEW

**Comment by davidwrighton** [comment]
@janvorli I don't have a comment on the correctness of this fix, as I'm not familiar with the debugger interaction here, but I see that just a few lines before this fix is a TODO-NewEH comment about `FastCallFinalize` and `COMToCLRDispatchHelperWithStack` Is that comment obsolete? Do we need to fix it for .NET 9?

**Comment by janvorli** [comment]
@davidwrighton I still need to investigate whether it is needed to do anything here. Thank you for noticing this / reminding me of this. None of the tests have ever hit a problem due to that and neither have customers reported any issues related to this. But it obviously doesn't mean it is fine as is, as a problem can be lurking waiting for someone to trigger it.

**Comment by janvorli** [comment]
Thanks @davidwrighton for the comment again. The finalizer case actually has similar problem as the one fixed by this PR, so I am going to add a fix to this PR too.
@jkotas, I've investigated possible fixes. One is to do the same thing as we did for the `CallDescrWorkerInternal`, but I've noticed that the `FastCallFinalizeWorker` exists only for x64 (besides x86). arm64 calls the finalizers via `CALL_MANAGED_METHOD_NORET`. I've tried to use the same for x64 and it also gets rid of the problem. I wonder - does the "fast call" still make sense or do you think that switching x64 to the same plan as other non-x86 architectures would be fine?

**Comment by jkotas** [comment]
> I wonder - does the "fast call" still make sense or do you think that switching x64 to the same plan as other non-x86 architectures would be fine?

I think it should be fine. `CALL_MANAGED_METHOD_NORET` reverse-FCall path is not that expensive. We did not have that path originally. You can try to build a small ad-hoc microbenchmark to see whether there is going to be significant measurable regression. If there is a significant measurable regression, it should be possible to fix by changing the polarity of the finalization loop. Run the finalization loop in managed code, and invoke the individual finalizers from managed code like we do it in naot: https://github.com/dotnet/runtime/blob/f8f450953d71dabe3fd01961c1d05b57c61e8faf/src/coreclr/nativeaot/Runtime.Base/src/System/Runtime/__Finalizer.cs#L55-L67

I would get rid of the x86 special path as well.

File: src/coreclr/vm/exceptionhandling.cpp [hunk 1]
```diff
@@ -8544,7 +8544,9 @@ extern "C" bool QCALLTYPE SfiNext(StackFrameIterator* pThis, uint* uExCollideCla
             _ASSERTE(retVal != SWA_FAILED);
             _ASSERTE(pThis->GetFrameState() != StackFrameIterator::SFITER_SKIPPED_FRAME_FUNCTION);
 
-            if (pThis->m_crawl.GetFrame() == FRAME_TOP)
+            pFrame = pThis->m_crawl.GetFrame();
+            // Check if there are any further managed frames on the stack, if not, the exception is unhandled.
+            if ((pFrame == FRAME_TOP) || (pFrame->GetVTablePtr() == DebuggerU2MCatchHandlerFrame::GetMethodFrameVPtr()))
```
Comment (jkotas) [inline src/coreclr/vm/exceptionhandling.cpp:8549]: Can `DebuggerU2MCatchHandlerFrame` be in the middle of the stack, with a bunch more managed frame behind i?

Comment (janvorli) [reply src/coreclr/vm/exceptionhandling.cpp:8549]: I have believed when I've searched some time ago that the only place where it is used is on top of secondary threads. But I have searched for it again to be 100% sure now and I have actually missed on in `DispatchInfo::InvokeMember`. So it seems that both this check and a similar one in the `NotifyExceptionPassStarted` should also check if that frame is also the topmost one.

**Comment by janvorli** [comment]
@jkotas I have added some instrumentation into the `FinalizerThread::FinalizeAllObjects`, summing the time it took to run the loop and number of items and printing it out every now and then. Then ran it with a test that creates tens of millions of finalizable objects with finalizer methods being empty.
The current `FastCallFinalizeWorker` way seems to be 15% faster than the CALL_MANAGED_METHOD_NORET way.

**Comment by jkotas** [comment]
I think it is acceptable regression. We can open a tracking issue to make the perf better by porting the naot scheme for finalizers.

**Comment by jkotas** [comment]
> We can open a tracking issue to make the perf better by porting the naot scheme for finalizers.

Actually, it is a nice cleanup. I have just done it: https://github.com/dotnet/runtime/pull/103501

File: src/coreclr/vm/exceptionhandling.cpp (hunk 1 above)
Comment (janvorli) [reply src/coreclr/vm/exceptionhandling.cpp]: @jkotas I have added check for the DebuggerU2MCatchHandlerFrame being the topmost one.

File: src/coreclr/vm/exceptionhandling.cpp [hunk 2]
```diff
@@ -8518,9 +8523,6 @@ extern "C" bool QCALLTYPE SfiNext(StackFrameIterator* pThis, uint* uExCollideCla
         }
         else
         {
-            // TODO-NewEH: Currently there are two other cases of internal VM->managed transitions. The FastCallFinalize and COMToCLRDispatchHelperWithStack
```
Comment (janvorli) [inline src/coreclr/vm/exceptionhandling.cpp:8521]: We don't propagate managed exceptions through the `COMToCLRDispatchHelper`, they are caught and transformed into HRESULT, so after the FastCallFinalize was removed by another PR, this comment is now obsolete.

**Review by jkotas — APPROVED** [review]

**Pull request merged**